# 程序流程控制
* 顺序结构
* 分支结构
* 循环结构

## 分支结构
* 根据条件控制哪个子代码块得到执行
* 使用if语句实现
* 单分支、双分支、多分支

单分支：当满足条件时，执行某个子代码块。条件可以是
* 关系表达式
* 逻辑表达式
* 算术表达式
* 其他可以转换为逻辑值的表达式

In [1]:
x = -65
y = x
if x < 0:
    y = -x
print(y)

65


In [2]:
a = [1,2,3,4]
x = len(a)
if(x>2):
    print('we have two more elements')
if(a):
    print('the list is not empty')

we have two more elements
the list is not empty


什么时候条件为假(False)？
* 数值0
* 空字符串''
* 空元组()
* 空列表[]
* 空字典或空集合{}
* bool值False
* None

都不行

负数是否为False? 虚数如2j是否为False？  都不是

前方有坑！注意语言的欺骗性。

In [3]:
if None:
    print('none')

双分支结构
* 使用if-else-构建
* 双分支结构都可以使用单分支结构完成

In [4]:
x=3
if(x>0):
    y=x
else:
    y=-x
y

3

双分支结构的单行表达式

In [5]:
y=x if(x>0) else -x
y

3

多分支结构

* 使用if-elif-...-elif-else实现
* 注意__elif__的拼写，含义为el(se)if
* 各个分支的条件采用顺序判断
* 第一个满足条件的分支被执行，并不是选择最合适的条件

In [6]:
mark = 85
if(mark>=90):
    grade = '优'
elif(mark>=80):
    grade = '良'
elif(mark>=70):
    grade = '中'
elif(mark>=60):
    grade = '及格'
else:
    grade ='不及格'
grade

'良'

In [4]:
# 找找这段代码中的错误，选中下列几行，使用ctrl+/切换注释和代码
y=2020
if(y%400 == 0):
    print('闰年')
elif(y%4 != 0):
    print('非闰年')
elif(y%100 == 0):
    print('非闰年')
else:
    print('是闰年')

是闰年


分支结构常见错误：
* if或者else子句漏了冒号':'
* 子代码块未缩进或缩进不一致
* 多个条件顺序错误
* 条件中的相等比较'=='错误写成了赋值'='

## 循环结构
* 循环结构用于重复执行某个子代码块
* 按照条件循环
* 指定次数循环

按照条件循环

* 不知道循环执行的次数，只知道循环结束的条件
* 使用while语句
* 需要确保循环条件能够达到
* 子代码块中应该含有能改变循环条件的语句

In [8]:
# 寻找调和级数前n项和超过给定目标的点
s=i=1;target=2.71
while(s<target):
    i+=1  # 这个条件必不可少！
    s+=1/i
print(i)

8


指定次数（范围）循环
* 知道循环执行的次数
* 或者用于遍历一个可迭代对象
* 使用for语句

In [9]:
# 计算调和系列前n项和
n=100; s=0
for i in range(1,n+1):
    s+=1/i
print(s)

5.187377517639621


In [10]:
for i in range(1,100):
    if (i%7==0): pass
    elif(i%10==7): pass
    elif(i//10==7): pass
    else: print(i,end=',')

1,2,3,4,5,6,8,9,10,11,12,13,15,16,18,19,20,22,23,24,25,26,29,30,31,32,33,34,36,38,39,40,41,43,44,45,46,48,50,51,52,53,54,55,58,59,60,61,62,64,65,66,68,69,80,81,82,83,85,86,88,89,90,92,93,94,95,96,99,

In [11]:
for i in range(1,100):
    if not(i%7==0 or i%10==7 or i//10==7):print(i,end=',')

1,2,3,4,5,6,8,9,10,11,12,13,15,16,18,19,20,22,23,24,25,26,29,30,31,32,33,34,36,38,39,40,41,43,44,45,46,48,50,51,52,53,54,55,58,59,60,61,62,64,65,66,68,69,80,81,82,83,85,86,88,89,90,92,93,94,95,96,99,

range对象：

* 迭代器对象，产生指定范围的整数序列
* 语法：range(start,stop[,step])
* 包含左边界start，不包含右边界stop
* start默认为0，步长step默认为1

In [12]:
s=0
for i in range(1,100,7):
    s+=i
print(s)

750


In [13]:
colors = ['red','blue','yellow']
for i in colors:
    i = i[0]
    print(i, end=' ')
colors

r b y 

['red', 'blue', 'yellow']

为了保证被遍历对象的安全性，防止被遍历对象被意外修改。遍历时对临时变量的修改并不会改变原始对象。如果需要修改原始对象，需要使用原始对象的索引访问。

In [14]:
colors = ['red','blue','yellow']
for index,value in enumerate(colors):
    colors[index]= value[0]
colors

['r', 'b', 'y']

break

* break语句用于结束循环
* 当有多个循环嵌套时，break只应用于最近的一层

In [15]:
# 判断一个数是否为素数
import math
x = 313; 
is_prime = True
n = int(math.sqrt(x))
for i in range(2,n+1): # 为什么stop参数设置为n+1,而不是n或者n+2?
    if(x%i==0):
        is_prime=False
        break
if(is_prime): print('是素数')
else: print('不是素数')

是素数


continue

* continue语句用于跳过循环的本轮执行
* continue只应用于最近的一层循环

In [16]:
s = 0
for i in range(1,100):
    if (i%7==0): continue
    elif(i%10==7): continue
    elif(i//10==7): continue
    else: s+=i
print(s)

3181


循环中的else子句

* for和while循环可以附带一个else子句
* 如果for和while语句没有被break中止，则else子句执行
* 可以将break语句理解为found，那么如果没找到，就执行else子句

In [17]:
# 判断一个数是否为素数
import math
x = 312; # x为大于1的整数
n = int(math.sqrt(x))
for i in range(2,n+1):
    if(x%i==0):
        print('不是素数')
        break
else: print('是素数')

不是素数


循环中else子句的注意事项

* else子句容易误解和出错，一般情形不建议使用
* else子句只有在break语句存在时才能正常工作

In [18]:
# 求立方根
def cubic_root(x):  # 定义一个立方根函数
    EPSILON = 1E-15
    guess = x
    while(abs(1-x/guess**3)>EPSILON):
        guess += (x/guess**2-guess)/3.0 # 使用牛顿迭代法
    return(guess)
print(cubic_root(15))

2.4662120743304703


In [19]:
x = 1234
EPSILON = 1E-15
guess = x
while(abs(1-x/guess**3)>EPSILON):
    guess += (x/guess**2-guess)/3.0 # 使用牛顿迭代法
    print(guess)  # 调试

822.6669367909237
548.445232306462
365.6315223710976
243.75742509965937
162.5118728092574
108.35682338383393
72.27291558231332
48.260692305809926
32.3504014217716
21.959972295453255
15.492944798515602
12.042294988959199
10.864648289390356
10.72777611879013
10.726014958033225
10.726014668827332
10.726014668827325


成对遍历

* 有时我们需要同时对两个相关的序列进行遍历
* 可以使用zip进行组装
* zip组装之后形成的是元组的列表

In [20]:
colors=['red','blue','yellow','black','purple','white','orange']
c_index = [1,3,2,5,6,7,4]
for i,j in zip(colors,c_index):
    print(i,j)

red 1
blue 3
yellow 2
black 5
purple 6
white 7
orange 4


练习：
<ol>
<li>若一个三位数的三个数字的立方之和恰好等于该数本身，则该数称为三位水仙花数。打印所有的三位水仙花数。</li>
<li>若一个五位数的五个数字的五次方的和恰好等于该数本身，则该数称为五位玫瑰花数。打印所有的玫瑰花数。</li>
<li>若一个数的真因子之和恰好等于该数本身，则该数称为完全数，例如6。找出1000以内的所有完全数。</li>
<li>根据党员月工资数额，求党员的月党费数值。</li>
</ol>

注：党员交纳党费的比例为：每月工资收入（税后）在3000元以下（含3000元）者，交纳月工资收入的0.5％；3000元以上至5000元（含5000元）者，交纳1％；5000元以上至10000元（含10000元）者，交纳1.5％；10000元以上者，交纳2％。


练习：根据个人的年收入，求其应当缴纳的个人所得税额。

首先，年应纳税所得额为年收入扣除6万元。
* 年应纳税所得额在3.6万元以内的，税率为第一级（3%）；
* 3.6万元至14.4万元部分适用税率第二级（10%）；
* 14.4万元至30万元部分适用税率第三级（20%）；
* 30万元至42万元部分适用税率第四级（25%）；
* 42万元至66万元部分适用税率第五级（30%）；
* 66万元至96万元部分适用税率第六级（35%）；
* 96万元以上部分适用税率第七级（45%）；